Update and comment BSI policy (Version 2024-01) (#4461)
randombit/botan: src/build-data/policy/bsi.txt

frodokem

# Table 2.5: Recommended parameters for ClassicMcEliece-KEM.
classic_mceliece

# === 3. Symmetric Encryption Schemes ===
# Table 3.1: Recommended block ciphers
aes

# Table 3.2: Recommended modes of operation for block ciphers
ccm
gcm
cbc
ctr

# Table 3.3: Recommended padding schemes for block ciphers
mode_pad # contains various paddings

# === 4. Hash Functions ===
# Table 4.1: Recommended hash functions
sha2_32
sha2_64
sha3

# === 5. Data Authentication ===
# Table 5.1: Recommended MAC schemes
cmac
hmac
kmac
gmac

# Table 5.3/5.5: Recommended signature algorithms.
# rsa (already enabled above)
dsa
ecdsa
ecgdsa
eckcdsa
xmss
hss_lms

# Table 5.4: Recommended padding schemes for the RSA signature algorithm.
emsa_pssr
iso9796 # DS 2 and 3

# === 8. Random Number Generators ===
auto_rng
hmac_drbg

# === Appendix and Others ===
# Table B.1: Recommended method for key derivation
sp800_56c # (Two-Step KDF)
hkdf

# B.1.3. Password-Based Key Derivation
argon2
argon2fmt

# Addition: ML-KEM, ML-DSA, and SLH-DSA
# We expect the BSI to approve the new FIPS (203,204,205) PQC algorithms
# in the upcoming TR (see Section 2.4.3 and Remark 5.5). We believe it is in

> # the BSI's interest to allow them here so applications may be mitigated to

Reviewer: ```suggestion
# the BSI's interest to allow them here so applications can migrate to
```

# post-quantum security as soon as possible.
ml_kem
ml_dsa
slh_dsa_shake
slh_dsa_sha2

# Optimization: PCurves
# To benefit from the speedup of pcurves, we activate all pcurve modules. We
# activate all curves regardless of the recommendation in the TR since they

> # would be accessible anyway in a worse generic implementation.

Reviewer 2: Keep in mind there is a code size imposition from each curve that's enabled, since the resulting object is specialized to that one curve. So enabling or disabling eg `pcurves_sm2p256v1` for sure does not prevent the curve from being available in any case, but having it be required in the build puts a minimum size on the resulting object. (That said this change is also requiring SL-DSA etc which are probably larger contributors to code size in aggregate)

Author: Good point. My main fear is that users may not know about pcurves and forget to activate them. Maybe it is sufficient for users optimizing for small binary sizes to apply the `--disable-modules` option. But yeah, I also think the size is quite small in comparison to the PQC algorithms.
